src/main/scala/com/yahoo/tensorflowonspark/TFModel.scala

```diff
@@ -0,0 +1,348 @@
+              byteBuf.get(byteArr)
+            case unsupportedType =>
+              Array()
+          }
+        }
+      }
+      result += Row(slices:_*)
+    }
+    result.toList
+  }
+
+  /**
+    * Transforms a Dataset of input data into an output DataFrame of predictions.
+    * Note: input columns are dropped during transform.
+    */
+  override def transform(dataset: Dataset[_]): DataFrame = {
+    val spark = dataset.sparkSession
+
+    val inputColumns = this.getInputMapping.keys.toSeq
+    val inputTensorNames = this.getInputMapping.values
+    val outputTensorNames = this.getOutputMapping.keys.toSeq
+
+    val inputDF = dataset.select(inputColumns.head, inputColumns.tail: _*)
+    val inputSchema = inputDF.schema
+    val outputSchema = transformSchema(inputSchema)
+
+    val outputRDD = inputDF.rdd.mapPartitions { iter: Iterator[Row] =>
+      if (TFModel.model == null || TFModel.modelDir != this.getModel) {
+        // load model into a per-executor singleton reference, if needed.
+        TFModel.modelDir = this.getModel
+        TFModel.model = SavedModelBundle.load(this.getModel, this.getTag)
+        TFModel.graph = TFModel.model.graph
+        TFModel.sess = TFModel.model.session
+      }
+
+      var results = List.empty[Row]
+      val groupedIter = iter.grouped(this.getBatchSize)
+      while (groupedIter.hasNext) {
+        val batch = groupedIter.next()
+
+        // get input batch of Rows and convert to list of input Tensors
+        val inputTensors = batch2tensors(batch, inputSchema)
+
+        var runner = TFModel.sess.runner()
+
+        // feed input tensors
+        for ((name, tensor) <- inputTensors) {
+          runner = runner.feed(this.getInputMapping(name), tensor)
+        }
+        // fetch output tensors
+        for (name <- outputTensorNames) {
+          runner = runner.fetch(name)
+        }
+
+        // run the graph
+        val outputTensors = runner.run()
+
+        assert(outputTensors.map(_.shape).map(s => if (s.isEmpty) 0L else s.apply(0)).distinct.size == 1,
+          "Cardinality of output tensors must match")
+
+        // convert the list of output Tensors to a batch of output Rows
+        val batchResults = tensors2batch(outputTensors)
+
+        // and add the batch of output Rows to the partition output
+        results ++= batchResults
+      }
+      results.iterator
+    }
+
+    spark.createDataFrame(outputRDD, outputSchema)
+  }
+
+  override def transformSchema(schema: StructType): StructType = {
+    if (TFModel.model == null || TFModel.modelDir != this.getModel) {
+      // load model into a driver singleton reference, if needed.
+      // Note: this implies that the driver memory must be sized similarly to the executors.
+      TFModel.modelDir = this.getModel
+      TFModel.model = SavedModelBundle.load(this.getModel, this.getTag)
+      TFModel.graph = TFModel.model.graph()
+      TFModel.sess = TFModel.model.session()
```

Comment:
why do we need TF session in driver?

Reply:
Likely not... was mostly ensuring the cached references were "complete".